/**
 * `InternalOAuthError` error.
 *
 * InternalOAuthError wraps errors generated by node-oauth.  By wrapping these
 * objects, error messages can be formatted in a manner that aids in debugging
 * OAuth issues.
 *
 * @constructor
 * @param {String} [message]
 * @param {Object|Error} [err]
 * @api public
 */
export class InternalOAuthError extends Error {
  oauthError: any;

  constructor(message?, err?) {
    super();

    this.name = this.constructor.name;
    this.message = message;
    this.oauthError = err;
    Error.captureStackTrace(this, this.constructor);
  }

  toString() {
    let m = this.name;
    if (this.message) {
      m += ': ' + this.message;
    }
    if (this.oauthError) {
      if (this.oauthError instanceof Error) {
        m = this.oauthError.toString();
      } else if (this.oauthError.statusCode && this.oauthError.data) {
        m +=
          ' (status: ' +
          this.oauthError.statusCode +
          ' data: ' +
          this.oauthError.data +
          ')';
      }
    }
    return m;
  }
}
